### DISCUSSION TITLE
Consider adding `refhook`?

### DISCUSSION THREAD

**Discussion by dipterix** [ropensci/targets#966] [Help] [answered]
Some references (external pointers, active bindings, environments) may break when saving R objects in separate rds, hence there is a `refhook` argument allowing users to re-link those references. 

Recently I ran into this issue where I have to manually re-link two R objects in all subsequent targets. Is there any similar mechanism like `refhook` in `targets`?

**Comment by wlandau** [comment, marked as answer]
I have no plans to allow refhooks because of the issues with `targets` internals/performance that would result, and because it is best practice to return simple objects that can be serialized and managed efficiently. Also, target storage formats like `"qs"` and `"feather"` are superior to RDS and do not support refhooks. The only reason `targets` uses RDS as the default is stability.

**Reply by dipterix** [reply]
Gotcha, so is it possible to write my own serializers/deserializers other than `qs` or `feather`? Or `targets::tar_format` can do that already?

In our case, we analyze >10GB data, which cannot be serialized via `qs` or `feather`: they are slow. Options such as `bigmemory` or `filearray` are considered as their speed could reach to GB/s. Currently I have to write to those external files, then generate "signatures" so targets can store, then restore the data in all targets that need them. Have to repeat myself over and over again.

I feel there must be some method that can avoid/simply this situation. For example, in my prepare script, I can register serializer/deserializer pairs with a string list this:

```r
library(targets)

targets::tar_format_register(format = "filearray", read = ..., write = ....)

...
tar_target(..., format = "filearray")

```

**Reply by dipterix** [reply]
NVM! I think you've already done it. :P

Now I'm greedy... is it possible to add `targets::tar_format_register` so the format can be "custom string"? I guess that would benefit Rmd users.

**Reply by wlandau** [reply]
It is possible to handle that case with the existing [tar_format()](https://docs.ropensci.org/targets/reference/tar_format.html).
